package cloudgene.mapred.jobs;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.HashMap;
import java.util.Map;

import cloudgene.mapred.jobs.state.JobState;
import cloudgene.mapred.jobs.workspace.IWorkspace;
import cloudgene.mapred.test.TestUtil;
import org.junit.jupiter.api.Test;

import cloudgene.mapred.TestApplication;
import cloudgene.mapred.core.User;
import cloudgene.mapred.database.dao.JobDao;
import cloudgene.mapred.database.dao.UserDao;
import cloudgene.mapred.jobs.workspace.WorkspaceFactory;
import cloudgene.mapred.util.config.Settings;
import cloudgene.mapred.wdl.WdlApp;
import cloudgene.mapred.wdl.WdlReader;
import genepi.io.FileUtil;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;

@MicronautTest
public class WrongWorkspaceTest {

	@Inject
	TestApplication application;

	@Inject
	WorkspaceFactory workspaceFactory;

	@Test
	public void testReturnTrueStep() throws Exception {
		WorkflowEngine engine = application.getWorkflowEngine();
		WdlApp app = WdlReader.loadAppFromFile("test-data/return-true.yaml");

		Map<String, String> inputs = new HashMap<>();
		inputs.put("input", "input-file");

		AbstractJob job = createJobFromWdl(app, inputs);
		engine.submit(job);

		TestUtil.waitForJob(engine, job);

		JobDao dao = new JobDao(application.getDatabase());
		AbstractJob jobFromDb = dao.findById(job.getId());

		assertEquals(JobState.FAILED, jobFromDb.getState());
		assertEquals(JobState.FAILED, job.getState());
	}

	public CloudgeneJob createJobFromWdl(WdlApp app, Map<String, String> inputs) throws Exception {
		UserDao userDao = new UserDao(application.getDatabase());
		User user = userDao.findByUsername("user");

		Settings settings = application.getSettings();

		String id = "test_" + System.currentTimeMillis();

		String localWorkspace = FileUtil.path("/gsfgdfgdf/vdadsadwa", id);
		FileUtil.createDirectory(localWorkspace);

		// setup workspace
		IWorkspace workspace = workspaceFactory.getDefault();
		workspace.setJob(id);
		workspace.setup();

		CloudgeneJob job = new CloudgeneJob(user, id, app, inputs);
		job.setId(id);
		job.setName(id);
		job.setWorkspace(workspace);
		job.setLocalWorkspace(localWorkspace);
		job.setSettings(settings);
		job.setApplication(app.getName() + " " + app.getVersion());
		job.setApplicationId(app.getId());

		return job;
	}
}
